### ISSUE TITLE
A way to get current PV/evaluation of the position

### ISSUE THREAD

**Issue by sphynx** [lcnr/rubot#3]
First, I'd like to thank you for the wonderful library! I'm using it to create an AI for a board game I'm playing.

However, I have a certain issue with the library: I don't see any way of getting the current evaluation (fitness) of the position, even though I would expect it to be one of the fields of the `Logger`.

Sometimes it's even clear from the logger that the analysis has completed, but it's still not clear whether the bot considers itself a loser or a winner :) Also, when the situation is not clear and analysis is not completed, usually chess engines allow users to see the PV ("principal variation", i.e. the current best path in the game tree) and its evaluation score:

https://www.chessprogramming.org/Principal_Variation

I tried to understand how to add PV or at least evaluation score in `rubot` code, but it looks non-trivial. Do you have any ideas or suggestions? 

Thank you!

**Comment by lcnr** [comment]
First of all thank you for this great issue :heart:

I haven't look at this repository in the last few months so I am not too familiar with this
code atm. Adding both the PV and the total evaluation score seems like a useful addition. :+1:

Will try to implement this in the following days. If you want to take a shot at this yourself,
I would also gladly help you out.

Afaict we can mostly (ab)use `BestAction` here, as it already stores the best path and its fitness for
performance reasons.
https://github.com/lcnr/rubot/blob/f3d49d19dbbf7cba2ceafaa770e9fdf453627ef6/src/alpha_beta.rs#L255-L259

This does not work for terminated paths though, which are stored without the taken path.
Not sure on how to best deal with this yet.
https://github.com/lcnr/rubot/blob/7bb25f4102ef1c0dae1eb03d6c0c0bb2a26595ae/src/alpha_beta.rs#L275-L281

**Comment by sphynx** [comment]
👍 Thanks for the prompt reply, it would be great to get this feature, I believe it will be invaluable for debugging bots!

For example, currently I have a situation when the bot's Logger claims to have analysed the tree down to `depth=9` (not completing the analysis though), then it chooses a move, then I make another move and it sees the loss in 3 moves! I.e. at the depth of 3, and the analysis is complete. I'm not quite sure how this can happen and whether it's a bug or a real thing, but I don't have many ways to debug such a situation and bot's decisions...

I've looked at your code, it looks quite neat and I like the API, especially an easy way of setting up terminating conditions, but the `alpha_beta.rs` looks complicated, not sure if I can change it and be sure that I haven't broken something :)

Last time I've coded alpha-beta pruning myself was something like 7-8 years ago, it was in Haskell (along with other relevant algorithms like nega-scout) here: 

https://github.com/sphynx/hamisado/blob/master/AI/Algorithms/My.hs

That code looks more or less simple, but I don't quite remember all the details right now...

It looks like part of the complexity in your code comes from the fact that you are combining alpha-beta with iterative deepening which I haven't done before. I'll take a closer look possibly combined with some reading about Alpha-Beta to refresh my mind on the topic. I even found a chapter on this in Russel and Norvig's "AI: a Modern Approach" book! :)

**Comment by sphynx** [comment]
BTW, if you are interested, I'm using `rubot` in my program for "Rokumon" game here: 

https://github.com/sphynx/rokumon/blob/master/src/ai.rs

**Comment by lcnr** [comment]
> For example, currently I have a situation when the bot's Logger claims to have analysed the tree down to depth=9 (not completing the analysis though), then it chooses a move, then I make another move and it sees the loss in 3 moves!

As long as you don't make a move which is not mentioned in `Game::actions`, this seems like a bug :sweat_smile:
I used `cargo fuzz` for a few hours before I released this crate and hoped this would be enough to get all missing bugs, but I probably underestimated the complexity of my alpha beta implementation. (or my fuzz harness has a bug (or both :shrug:))

IIRC I wanted to optimize based on partially terminating paths (which is probably also the cause of your bug), so each step of iterative deepening can skip all paths which are guaranteed to be worse than the best already terminated one.

> but the alpha_beta.rs looks complicated, not sure if I can change it and be sure that I haven't broken something :)

Don't worry about that :sparkles: Either the current tests catch it or we have new tests we can add

**Comment by lcnr** [comment]
> but the alpha_beta.rs looks complicated

After looking at it a bit more, I have to say it's actually horrible :laughing:
Will first try to clean this up before working on this issue.

**Comment by sphynx** [comment]
> After looking at it a bit more, I have to say it's actually horrible 😆

I've looked at it once more too, but it was too hard, I couldn't follow the general control flow and the interaction between all those data types: Terminated, Minimax, Branch, RateAction and so on. It all kinda makes sense from the very high view, but when I started to dig deeper I was feeling lost, since sometimes the assumptions are not clear, comments assume some internal knowledge and the algorithm looks quite different to like 10 lines in Wikipedia (I still don't quite understand *why* it's so different: is that because of efficiency reasons? Or it's not pure AlphaBeta, but also some optimisations?). 

It looks like a document detailing the overall algorithm and internal data structures may be helpful for contributors.

> Will first try to clean this up before working on this issue.

Looking forward to that! Thank you for treating this issue seriously.

**Comment by lcnr** [comment]
btw, how do you expect this information to be available?

I think it's easiest to give them to the `RunCondition` which can then decide what to do with that information. So both the expected path and the expected fitness will be available as a method of `Logger`.

**Comment by sphynx** [comment]
Yes, either from `Logger` or as a result of `bot.select()` would be both fine for my needs.

Do you think it will have some performance cost to always return PV and fitness from `select`?

**Comment by lcnr** [comment]
> Do you think it will have some performance cost to always return PV and fitness from select?

I think it should be fairly negligible :laughing:

**Comment by lcnr** [comment]
Added `Bot::detailed_select` on master, with returns a struct containing the expected fitness
and the expected path.

I also added https://crates.io/crates/tapir as a dependency because dealing with paths was annoying :sweat_smile: 

I am not certain if this is a good API so I did not yet release a new version.

^ *edit*: I also want better benchmarks before I release this, I don't have enough atm and the current ones take too long. It is also a pain because there are changes which improve nearly all test cases by a decent amount but triple the steps require for another test :crying_cat_face:

**Comment by sphynx** [comment]
Great! I'll just use dev version of `rubot` for now, will let you know how it goes. Don't worry about releasing it quickly, I'm in no rush.

**Comment by sphynx** [comment]
I've updated to the latest version from GitHub. 

Currently it looks like in one position I'm getting very different number of steps from one run to another, is that expected? Is there any randomness in bot decisions or data structures which can give different iteration order from run to run (say `HashMap` or `HashSet` which use a randomised hasher by default)?

```
Run 1:
Running AI with duration 15s...
AI log: steps: 336118, depth: 9, completed: true, duration: 11.080560224s
AI recommends: move R6 from <2, 0, -2> to <0, 0, 0>

Run 2:
Running AI with duration 15s...
AI log: steps: 419558, depth: 9, completed: true, duration: 12.989010761s
AI recommends: move R6 from <2, 0, -2> to <0, 0, 0>

Run 3:
Running AI with duration 15s...
AI log: steps: 340665, depth: 9, completed: true, duration: 11.212721688s
AI recommends: move R6 from <2, 0, -2> to <0, 0, 0>
```

**Comment by sphynx** [comment]
Sometimes I can see that depth reported in the log is 8, while PV contains 9 actions, is that expected?

**Comment by lcnr** [comment]
> Sometimes I can see that depth reported in the log is 8, while PV contains 9 actions, is that expected?

Yes, as depth 0 means that we sort by `state.look_ahead(action, self.player)`. The reason we sometimes have a depth of 8 and and PV with 8 actions is probably because the depth was not finished, causing us to use an evaluation of the previous depth.

**Comment by lcnr** [comment]
I actively tried to avoid any randomness here, so this should not happen as long as the computation has finished.

If you use `Duration` as a `RunCondition`, you can expect differences though.
Do these changes persist when using `ToCompletion`?

**Comment by sphynx** [comment]
> I actively tried to avoid any randomness here, so this should not happen as long as the computation has finished.

You can see that in all three cases, `completed = true`, I assume that this means that it has completed.

> If you use `Duration` as a `RunCondition`, you can expect differences though.

Note that I used 15s Duration and it had completed long before that (i.e. a couple of seconds earlier). Is that still expected to lead to non-determinism? I would expect the non-determinism only in the case when it's not completed (since we can stop the computation in different points when duration elapsed)...

> Do these changes persist when using `ToCompletion`?

No, it's stable then:

```
> complete
Running AI until completion...
AI log: steps: 434946, depth: 9, completed: true, duration: 13.867996018s
AI evaluation: -2147483648
PV:
1: move R6 from R2C3 to R2C1, eval: -2
2: place B3 at R2C3, eval: -4
3: move R6 from R2C1 to R1C1, eval: -4
4: place B5 at R2C4, eval: -2147483648

AI recommends: move R6 from R2C3 to R2C1

> complete
Running AI until completion...
AI log: steps: 434946, depth: 9, completed: true, duration: 13.27320522s
AI evaluation: -2147483648
PV:
1: move R6 from R2C3 to R2C1, eval: -2
2: place B3 at R2C3, eval: -4
3: move R6 from R2C1 to R1C1, eval: -4
4: place B5 at R2C4, eval: -2147483648

AI recommends: move R6 from R2C3 to R2C1
```

**Comment by sphynx** [comment]
Ah, no, sorry. It's different between different program runs even when run to completion, but the same in one run. 

I think the problem might be in the fact that you use unstable sort: `game_states.sort_unstable_by(|(_, _, a), (_, _, b)| a.cmp(b));` which is based on a randomised algorithm, so game states with the same fitness may be traversed in different order.

**Comment by lcnr** [comment]
uff, don't know why that's there, I now use an ordinary sort instead.

Can you try again it now? :sweat_smile:

**Comment by sphynx** [comment]
Haha, it looks like on top of that I also used a `HashMap` in one of my structures and iterating over that might lead to a different result in each run! 😃 Changed that to `BTreeMap` which gives consistent ordering.

Now it works fine. Actually now it gives consistent results between runs with both `sort_unstable_by` and `sort_by` in your code (however the number of steps are different between those two versions of code). But still I'm not sure if using unstable sort is a good idea there, but it's your call.

**Comment by lcnr** [comment]
:sweat_smile: 

I think it's mostly irrelevant Most of the time there are less than 20 possible actions, meaning that both `sort` and `sort_unstable` should just default to insertion sort.

**Comment by sphynx** [comment]
Do you mind changing the order to a more natural for PV? I.e. from first move to the last? Otherwise you just push this work to the client :)

**Comment by lcnr** [comment]
Ok

**Comment by sphynx** [comment]
I have one more question regarding this, even though not directly related to this issue.
 
When the bot is run to completion and discovers that he loses with optimal play, it starts making moves which are quite rubbish (i.e. have lower immediate fitness) and lead to the loss faster than if played otherwise. 

For example, here we run the bot with depth=8, it still doesn't see that it's losing, so it proposes a decent move with immediate fitness=0:

```
> depth 8
Running AI with depth Depth(8)...
AI log: steps: 88587, depth: 8, completed: false, duration: 2.858292202s
AI evaluation: -1
PV:
1: place r2 at r2c1, eval: 0
2: move b1 from r1c2 to r2c2, eval: 0
3: place r2 at r2c4, eval: 1
4: place b3 at r1c1, eval: 0
5: move r2 from r2c1 to r1c1, eval: 2
6: fight at r1c1, eval: -1
7: move r2 from r2c4 to r1c1, eval: 1
8: fight at r1c1, eval: -2
9: place r4 at r2c4, eval: -1

AI recommends: place r2 at r2c1
```

But now I re-run it until completion, it sees that it's losing and gives a much shorter PV and selected move which has worse immediate fitness: -2:

```
> complete
Running AI until completion...
AI log: steps: 365838, depth: 9, completed: true, duration: 11.711602266s
AI evaluation: -2147483648
PV:
1: move r6 from r2c3 to r2c1, eval: -2
2: place b3 at r2c3, eval: -4
3: move r6 from r2c1 to r1c3, eval: -4
4: place b5 at r2c1, eval: -2147483648

AI recommends: move r6 from r2c3 to r2c1
```

So, with this move, AI is going to lose very quickly: after just 3 more moves.

However, let's see what happens if we play that move proposed before (with analysis for depth=8):

```
> place r2 at r2c1
Played move: place r2 at r2c1

> complete
Running AI until completion...
AI log: steps: 119527, depth: 8, completed: true, duration: 3.823930201s
AI evaluation: 2147483647
PV:
1: move b1 from r1c2 to r2c2, eval: 0
2: move r6 from r2c3 to r2c1, eval: -1
3: place b3 at r2c3, eval: -3
4: fight at r2c1, eval: -4
5: place b5 at r2c4, eval: -2147483648
```

Now you can see that the bot is actually only able to beat that move after 5 more moves (which looks better than just 3 moves in the previous scenario).

**To summarise**: the problem here is when the bot sees that it loses, it doesn't try to play optimal moves anymore. Or at least it doesn't play optimal moves in terms of playing as long as possible before actually losing. Do you know if this is expected? Or is it a general problem of alpha-beta?

Do you know how PV is supposed to look like when the bot knows that it loses? What criteria are then used to choose the best line in this situation?

Thanks!

**Comment by lcnr** [comment]
> Or at least it doesn't play optimal moves in terms of playing as long as possible before actually losing.

Yes, that's expected, the bot does not care about the total amount of moves at all. One way
you may be able to fix this is by increasing the lower bound by 1 for each turn taken. This should cause the bot to prioritize staying alive once it is certain to lose.

> What criteria are then used to choose the best line in this situation?

We only replace the desired move if `fitness > old_best` so the first fully terminated losing move should always be chosen atm.

There is a way to always force this behavior in rubot itself, I guess this would always be expected?
Can you think of a situation where we want the bot to fail quickly?

**Comment by sphynx** [comment]
> Yes, that's expected, the bot does not care about the total amount of moves at all. 

Makes sense.

> One way you may be able to fix this is by increasing the lower bound by 1 for each turn taken. 

Hm, I didn't quite understand this approach. Do you mean changing `is_lower_bound` function in trait implementation? I.e. something like this:

```rust
impl rubot::Game for Game {
  ...
  fn is_lower_bound(&self, fitness: Self::Fitness, _player: Self::Player) -> bool {
    // before: fitness == i32::MIN
    // after:
    fitness <= i32::MIN + self.num_of_moves_played()    
  }
}
```

but this doesn't make much sense to me. Perhaps you meant something else...

> We only replace the desired move if fitness > old_best so the first fully terminated losing move should always be chosen atm.

Makes sense. So, if we take this to the extreme and the very first child of the root node has `-inf` fitness, the path is terminated and that child will be kept as the best move so far. And then, if all the next nodes are not able to improve on that, at the end we will just return that very first child and will just lose in 1 move.

> There is a way to always force this behavior in rubot itself, I guess this would always be expected?
Can you think of a situation where we want the bot to fail quickly?

I think so, there is not much to lose for the bot. I.e. continuing to play in this situation may only decrease the chance of losing from 100% to something smaller due to opponent's mistake. And the bot is much less ethically bound to resign than human players :)


BTW, on a completely different topic, I finally understood what this data structure in the code represents:

```
enum Branch<T: Game> {
    /// `actual_fitness <= fitness`
    Worse(T::Fitness),
    /// `actual_fitness >= fitness`
    Better(T::Fitness),
    /// `actual_fitness == fitness`
    Equal(T::Fitness),
}
```

It just represents beta-cutoff (aka Better, which sounds very similar to "beta" :D ), alpha-cutoff and no-cutoff. It might worth noting that somewhere in the doc comment for further readers :)

Overall, I think the code has much more helpful comments now and ALPHA_BETA.md is also quite readable and helpful, so more of the code now makes sense to me!

**Comment by lcnr** [comment]
> One way you may be able to fix this is by increasing the lower bound by 1 for each turn taken.
```rust
impl rubot::Game for Game {
  fn execute(&mut self, action: &Self::Action, player: Self::Player) -> Self::Fitness {
       let fitness = // the previous method body;
       if fitness == i32::MIN {
          // Still less than `lower_bound`, but a later move is slightly better.
          fitness + self.num_of_moves_played()  
       } else {
          fitness
        }
  }
  ...
  fn is_lower_bound(&self, fitness: Self::Fitness, _player: Self::Player) -> bool {
    // Depending on how many turns are possible
    fitness <= i32::MIN + 10000
  }
}
```
I just updated `rubot` itself just now, so don't use this :laughing:  By changing this behavior we may even speed things up in some cases :sparkles:

Still fuzzing, so there still may be an unexpected panic/wrong behavior here.

> Makes sense. So, if we take this to the extreme and the very first child of the root node has -inf fitness, the path is terminated and that child will be kept as the best move so far. And then, if all the next nodes are not able to improve on that, at the end we will just return that very first child and will just lose in 1 move.

:+1: 

> I think so, there is not much to lose for the bot. I.e. continuing to play in this situation may only decrease the chance of losing from 100% to something smaller due to opponent's mistake. And the bot is much less ethically bound to resign than human players :)

Makes sense, and in case a bot should surrender one can do this manually be checking the expected fitness for more details.

> Overall, I think the code has much more helpful comments now and ALPHA_BETA.md is also quite readable and helpful, so more of the code now makes sense to me!

Nice :tada:

**Comment by sphynx** [comment]
Ah, I see, that code makes sense. 

> I just updated rubot itself just now, so don't use this 😆 

Wow, what a quick turnaround! I'll automatically test it too, since my `Cargo.toml` now has an override which directly points to the latest `rubot` commit in master :D

> By changing this behavior we may even speed things up in some cases ✨

Are you still running your chess bot on Lichess? It would be nice to see if those improvements can lead to a higher rating!

**Comment by lcnr** [comment]
> Are you still running your chess bot on Lichess? It would be nice to see if those improvements can lead to a higher rating!

I don't, for whatever reason it died every few days :thinking: May try and get that one back running,
but am not that motivated to do so rn :/

**Comment by sphynx** [comment]
I've tested your change in that situation I was describing [above](https://github.com/lcnr/rubot/issues/3#issuecomment-629716577), now the bot works great: it returns the same reasonably looking move even with `-inf` evaluation discovered with deeper depths.

I think we can now resolve this issue, if your fuzzing shows no problems. 

Thanks a lot for your work, amazingly quick turnaround and patience while answering my questions! I'm looking forward to `rubot-0.3`!

**Comment by sphynx** [comment]
Hi @lcnr! Just wanted to ask whether you are going to publish the changes as a new version? I'm still depending on a Github dev version and it would be nice to update my Cargo.toml to use 0.3 as a dependency.

Thank you!

**Comment by lcnr** [comment]
I want to, but I am currently not yet satisfied with the API of `alpha_beta::Bot`... I don't really like `detailed_select` here.

Considering that you are the biggest "customer" (i.e. the only one I know of), I can go ahead and do some cleanup and then publish a new version this weekend if you are satisfied with the current API.

**Comment by sphynx** [comment]
Yeah, I'm fine with the current API, i.e. I can use it and get all information I need, so it works. But of course, as a library designer you may have stricter views on the design.

It would be great if you can release 0.3 over the weekend!

**Comment by sphynx** [comment]
BTW, you may be interested that I am using a version of `rubot` compiled to WebAssembly in a web project.

It's mostly working fine out of the box, the only thing I needed to change was related to timing durations. Since `std::time` is not supported in browser, in order to use Duration-based terminating conditions I had to implement my own condition which gets time information from the browser using WebAPI (with standard `web-sys` crate).

Another thing that I couldn't use `Logger` out of the box, since it also depends on time by default.

If you are interested, we can try to factor in this functionality and make `rubot` more WebAssembly-friendly, so that it can be used in web game bots.

**Comment by lcnr** [comment]
Added some docs to `detailed_select` and published version 0.3.0.

I do think having good wasm support is a very good thing. I don't have too much time rn and
am not too familiar with the current state of rust in this area though.
So while I will very gladly review changes/give feedback, I won't work on this myself, at least not
in the near future.

**Comment by sphynx** [comment]
> Added some docs to detailed_select and published version 0.3.0.

Great, thanks!

> I do think having good wasm support is a very good thing. I don't have too much time rn and
am not too familiar with the current state of rust in this area though.
So while I will very gladly review changes/give feedback, I won't work on this myself, at least not
in the near future.

Sure, I mostly meant that if there is interest I can try to incorporate my changes and send a pull request. Now I see that there is interest, so I'll try to contribute.
